### ISSUE TITLE
Name conflicts with State monad

### ISSUE THREAD

**Issue by danidiaz** [winterland1989/data-has#1]
The names "get" and "modify" collide with functions from [Control.Monad.Trans.State](http://hackage.haskell.org/package/transformers-0.5.2.0/docs/Control-Monad-Trans-State-Strict.html#v:get).

An alternative proposal: the `Has a t` typeclass could have a single member `part :: Lens t a`. The lens could be defined without depending on any particular lens library.

`get` and `modify` could be separate functions that perhaps could even be in a separate module (`Data.Has.Simple`?)

The [`view`](http://hackage.haskell.org/package/lens-4.15.1/docs/Control-Lens-Getter.html#v:view) function from `lens` works in a `MonadReader` context (sadly, the one from `microlens` doesn't)  so we could write things like `view part` instead of `asks get`.

**Comment by winterland1989** [comment]
Hi Daniel!

I actually have considered about lens in the first place, see [my original mail](https://mail.haskell.org/pipermail/haskell-cafe/2016-November/125587.html). But i finally decided go with get&set because lens and get&set can be defined by each other, the real question is which is more suitable for this use case:

+ Lens is intend to manipulate nested data structures, which is rare is this case.
+ Even if you have a nested tuple here, you would still use get or set because you have to annotated type for every middle step, because `Has` only provide typed field selection.
+ get&set is easier to implement(with some editor tricky), and much beginner friendly.
+ You don't have to ask user for an extra dependency.

So in short i think provide get&set is the right thing to do for 99% of the time. But i think the naming collision is indeed an accident. While you can qualify `get/modify` with `Has`, I think it's OK to change names. What about `getter` and `setter`?

**Comment by winterland1989** [comment]
Rename to `getter/modifier` in c05e5fd4ad70135cd2eadca42567afeed8a438c1.

**Comment by danidiaz** [comment]
Thanks for the link, I have now read the thread.

I agree that having to define a lens would be less beginner-friendly.

That said, it would not be necessary to ask the user for an extra dependency, because lenses can be defined independently. For example:

    -- internal definitions, non-exported
    type Lens s t a b = forall f. Functor f => (a -> f b) -> s -> f t
    type Lens' s a = Lens s s a a
    view :: s -> Lens' s a -> a
    view s l = getConst (l Const s)
    over :: Lens' s a -> (a -> a) -> s -> s
    over l f = runIdentity . l (Identity . f)

    -- exported
    firstofofur :: Lens' (a, b, c, d) a
    firstofofur f (a, b, c, d) = fmap (\a' -> (a',b,c,d)) (f a)

    get :: (a, b, c, d) -> a
    get t = view t firstofofur

    modify :: (a -> a) -> (a, b, c, d) -> (a, b, c, d)
    modify f = over firstofofur f

And the exported `firstofofur` lens would be compatible with a standard lens package.

**Comment by winterland1989** [comment]
OK, but since the use case is very specific, for simplicity i'll leave it as it now. If you really need a lens you can define it with `getter` and `modifier`.

**Comment by winterland1989** [comment]
We got `hasLens` now，closing
